### PR TITLE
Add mocks package for unit testing

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
This includes a mock net.Conn and various helper funcs for creating mock
response frames.
Moved frame parsing out of conn.go and into frames package so that the
mock infrastructure can use it.
Fixed a bug in link uncovered by new unit tests.
Rewrote receiver tests using the new framework.

### CODE REVIEW

File: receiver_test.go [hunk 1]
```diff
@@ -2,112 +2,161 @@ package amqp
...
-		t.Errorf("Receive() error = %v", err)
-	}
-	if len(r.dispositions) != 0 {
-		t.Errorf("it is up to the message handler to settle messages")
+func TestReceive_ModeSecond(t *testing.T) {
+	const linkName = "test"
+	const linkHandle = 0
+	deliveryID := uint32(1)
+	responder := func(req frames.FrameBody) ([]byte, error) {
+		switch ff := req.(type) {
```
Comment (richardpark-msft) [inline receiver_test.go:96]: This looks pretty great!

File: link.go [hunk 2]
```diff
@@ -217,6 +217,9 @@ func attachLink(s *Session, r *Receiver, opts []LinkOption) (*link, error) {
 	}
 
 	if isReceiver {
+		if l.Source == nil {
```
Comment (richardpark-msft) [inline link.go:220]: Ah, this would be the bug you mentioned in the PR message?

**Review by richardpark-msft — APPROVED** [review]
Looks good, glad to see more stuff go into internal.

And the mock connection stuff is looking great, should really help.

File: link.go (hunk 2 above)
Comment (jhendrixMSFT) [reply link.go:220]: Yeah, I improved on the fix in a later commit and applied the same to `l.Target`

**Pull request merged**